legion_experiment/src/query/view.rs

```diff
@@ -0,0 +1,349 @@
+use crate::borrow::{Exclusive, Shared};
+use crate::storage::archetype::Archetype;
+use crate::storage::chunk::Chunk;
+use crate::storage::components::{Component, ComponentTypeId};
+use derivative::Derivative;
+use std::any::TypeId;
+use std::marker::PhantomData;
+use std::slice::{Iter, IterMut};
+
+pub trait View<'data> {
+    type Elements: Fetch<'data>;
+
+    unsafe fn fetch(archetype: &'data Archetype, chunk: &'data Chunk) -> Self::Elements;
+
+    fn validate();
+
+    /// Determines if the view reads the specified data type.
+    fn reads<T: Component>() -> bool;
+
+    /// Determines if the view writes to the specified data type.
+    fn writes<T: Component>() -> bool;
+}
+
+pub trait Fetch<'a> {
+    type Iter: Iterator + 'a;
+    unsafe fn iter_mut(&mut self) -> Self::Iter;
+    unsafe fn find_components<T: Component>(&self) -> Option<&[T]>;
+    unsafe fn find_components_mut<T: Component>(&mut self) -> Option<&mut [T]>;
+}
+
+pub trait ReadOnlyFetch<'a>: Fetch<'a> + ReadOnly {
```

Comment:
`ReadOnly` and `ReadOnlyFetch` are both marker traits which indicate views (and their fetches) which do not perform any mutable accesses. It is used as a constraint in e.g. `Query::iter`, so that you can run a query with only `&World` with safe code, but only when you are not writing to any components.

These traits should probably be marked `unsafe`, actually (much like `Send` and `Sync` for similar reasons).